validator/sawtooth_validator/journal/validation_rule_enforcer.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from sawtooth_validator.protobuf.transaction_pb2 import TransactionHeader
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ValidationRuleEnforcer(object):
+    def __init__(self, setting_view_factory):
+        self._settings_view_factory = setting_view_factory
+
+    def validate(self, block, state_root):
+        """
+        Retrieve the validation rules stored in state and check that the
+        block does not violate any of those rules. These rules include:
+
+            NofX: Only N of transaction type X may be included in a block.
+            XatY: A transaction of type X must be in the block at position Y.
+            local: A transaction must be signed by the same key as the block.
+
+        If any setting stored in statem, does not match the required format for
+        that rule, the rule will be ignored.
+
+        """
+        settings_view = self._settings_view_factory.create_settings_view(
+            state_root)
+        rules = settings_view.get_setting(
+            "sawtooth.validator.block_validation_rules")
+        if rules is None:
+            return True
+
+        transactions = []
+        for batch in block.batches:
+            transactions += batch.transactions
+
+        expected_signer = block.header.signer_pubkey
+
+        rules = rules.split(";")
+        valid = True
+        for rule in rules:
+            try:
+                rule_type, arguments = rule.split(":")
+            except ValueError:
+                LOGGER.warning("Validation Rule Ignored, not in the correct "
+                               "format: %s",
+                               rule)
+                continue
+
+            rule_type = rule_type.strip()
+            if rule_type == "NofX":
+                valid = self._do_nofx(transactions, arguments)
+
+            elif rule_type == "XatY":
+                valid = self._do_xaty(transactions, arguments)
+
+            elif rule_type == "local":
+                valid = self._do_local(
+                    transactions, expected_signer, arguments)
+
+            if not valid:
+                return False
+
+        return valid
+
+    def _do_nofx(self, transactions, arguments):
```

Comment:
Also, a doc string of what this rule does would be very helpful.  Hard to review without knowing what these are supposed to be evaluating.

Reply:
Added